### ISSUE TITLE
federation: verify and add tests for replicaset preferences annotations

### ISSUE THREAD

**Issue by nikhiljindal** [kubernetes/kubernetes#33312]
Found while writing the federation replicaset guide:

Steps to repro:
- Register a cluster C1 for federation.
- Create a federation replicaset with X replicas.
- Observe that a replicaset is created in the cluster C1 with X replicas.
- Now register another cluster C2 for federation.

Expected:
- The replicaset is also created in cluster C2 with X/2 replicas. Replica set in cluster C1 is scaled down from X to X/2 replicas.

Actual:
- Replicaset not created in cluster C2. Replicaset in cluster C1 continues to have X replicas.

cc @kubernetes/sig-cluster-federation

**Comment by nikhiljindal** [comment]
I see now that we do not have any e2e test for this case (register a cluster after a replicaset is created)

**Comment by nikhiljindal** [comment]
cc @kshafiee @jianhuiz

**Comment by nikhiljindal** [comment]
Reading the description of [rebalance](https://github.com/kubernetes/kubernetes/blob/cf4ea938c882d5f86852994807b18959458033d5/federation/apis/federation/types.go#L121) field in FederatedReplicaSetPreferences, it looks like rebalancing is expected not to happen unless rebalance is set to true.

I tried setting rebalance = true in the annotation on the federated replicaset before creating it and it still didnt rebalance.

**Comment by jianhuiz** [comment]
you need to specify it like this
```
  annotations:
    federation.kubernetes.io/replica-set-preferences: >
      {
        "rebalance": true,
        "clusters": {
          "*": {
            "weight": 1
          }
        }
      }
```

**Comment by nikhiljindal** [comment]
Yes I realised I wasnt doing it right the first time.
It seemed as if all was fine, but then I saw controller manager logs and saw errors in parsing my annotation. Ideally my create request should fail if the annotation is incorrect.
I got it right finally (no annotation parsing errors), but my replicaset still didnt get rebalanced when I added a cluster.
We should really add an e2e test for this.

This is not a release blocker since the annotation is alpha.

**Comment by jianhuiz** [comment]
@nikhiljindal I wasn't able to reproduce this, could you upload the controller logs and the replicaset file you use?

**Comment by nikhiljindal** [comment]
The annotation I had tried I think was:

```
annotations:
    federation.kubernetes.io/replica-set-preferences: >
      {
        "rebalance": true,
        }
      }
```

We should add a test case for that.
Can you or @mwielgus take that up?

**Comment by jianhuiz** [comment]
not specifying weight would implies 0 for all clusters which results in no replicaset would be created in the underlying clusters.

**Comment by mwielgus** [comment]
Yep, i can take it.

**Comment by nikhiljindal** [comment]
Another replicaset issue reported on slack: we do not set the "ready" field in replicaset status: https://kubernetes.slack.com/archives/sig-federation/p1478597937000904

**Comment by jianhuiz** [comment]
and AvailableReplicas, too, which are introduced later to fed replicaset. I'll add them tomorrow.
#29481 #32771

**Comment by csbell** [comment]
@mwielgus, I can take this one.

**Comment by henriquetruta** [comment]
@nikhiljindal @jianhuiz didn't work for me. Used the annotation as you suggested and the replicas weren't rebalanced

**Comment by nikhiljindal** [comment]
Updated the issue title to be generic to add tests for replica set preferences.
@henriquetruta is taking a first cut at https://github.com/kubernetes/kubernetes/pull/45840

**Comment by kustuntas** [comment]
@nikhiljindal Has the rebalancing issue been fixed yet?

**Comment by nikhiljindal** [comment]
@kustuntas I havent tried it again but it doesnt seem to be based on updates on this bug.

cc @perotinus

**Comment by quinton-hoole** [comment]
Unassigning mwielgus, who I'm pretty sure is not working on this.  
Sounds like we have both a missing test (addressed in #45840), and possibly a bug.
We'll need to resolve this to GA federation.

**Comment by irfanurrehman** [comment]
This issue was labelled only for sig/multicluster and is thus moved over to kubernetes/federation#74.
If this does not seem to be right, please reopen this and notify us @kubernetes/sig-multicluster-misc.
/close
